File: python/bc-ur/src/bc_ur/multipart_encoder.py
"""Multipart UR encoder using fountain codes."""

from __future__ import annotations

from ._fountain_encoder import FountainEncoder
from .bytewords import BytewordsStyle, encode as bw_encode
from .ur import UR


class MultipartEncoder:
    """Encodes a UR as a stream of multipart UR strings using fountain codes."""

    __slots__ = ("_encoder", "_ur_type")

    def __init__(self, ur: UR, max_fragment_len: int) -> None:
        data = ur.cbor.to_cbor_data()
        self._encoder = FountainEncoder(data, max_fragment_len)
        self._ur_type = ur.ur_type_str

    def next_part(self) -> str:
        """Emit the next multipart UR string."""
        part = self._encoder.next_part()
        body = bw_encode(part.to_cbor(), BytewordsStyle.MINIMAL)
        return f"ur:{self._ur_type}/{part.sequence_id()}/{body}"

    @property
    def current_index(self) -> int:
        """The current sequence number (1-based, incremented after each part)."""
        return self._encoder.current_sequence

    @property
    def parts_count(self) -> int:
        """The number of original fragments (minimum parts for full coverage)."""
        return self._encoder.fragment_count

File: typescript/bc-ur/src/multipart-encoder.ts
import { Style, encode as encodeBytewords } from "./bytewords.js";
import { URError } from "./error.js";
import { FountainEncoder } from "./internal/fountain.js";
import { UR } from "./ur.js";
import { messageFromUnknown } from "./utils.js";

/**
 * Multipart UR encoder backed by fountain coding.
 *
 * Splits a UR payload into fragments that can be transmitted as a sequence
 * of UR strings. Mixed (XOR) fragments enable the decoder to reconstruct
 * the original message even if some parts are missed.
 */
export class MultipartEncoder {
  readonly #encoder: FountainEncoder;
  readonly #urType: string;

  constructor(ur: UR, maxFragmentLen: number) {
    this.#urType = ur.type;

    try {
      this.#encoder = new FountainEncoder(ur.cbor().toData(), maxFragmentLen);
    } catch (error) {
      throw URError.ur(messageFromUnknown(error));
    }
  }

  /**
   * Returns the next multipart UR part string.
   */
  nextPart(): string {
    try {
      const part = this.#encoder.nextPart();
      const body = encodeBytewords(part.cbor(), Style.MINIMAL);
      return `ur:${this.#urType}/${part.sequenceId()}/${body}`;
    } catch (error) {
      throw URError.ur(messageFromUnknown(error));
    }
  }

  /**
   * Returns the current emitted part index (starts at 0).
   */
  get currentIndex(): number {
    return this.#encoder.currentSequence();
  }

  /**
   * Returns the total fragment count for the source message.
   */
  get partCount(): number {
    return this.#encoder.fragmentCount();
  }
}
